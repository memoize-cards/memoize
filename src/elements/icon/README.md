# Icon

The `Icon` component is a versatile and customizable icon library that provides a collection of SVG icons for use in your web applications. The library includes various icons for common actions and UI elements, such as arrows, symbols, and indicators. The `Icon` component allows you to easily render these icons using simple JSX syntax.

### Usage

1. Import the `Icon` component into your component or other frontend files:

```js
import icon from '@elements/icon'
```

2. Render the desired `icon` using the Icon component:

```js
const MyComponent = () => {
  return (
    <div>
      <h1>My Web Application</h1>
      <p>Click the arrow to navigate back.</p>
      <Icon.CaretLeft />
    </div>
  )
}
```

In the above example, we are rendering the `CaretLeft` icon from the `Icon` component library using the JSX syntax `<Icon.CaretLeft />`. This will render the SVG icon representing a left-pointing caret arrow.

### Contributing

Contributions are welcome! If you find any issues or have suggestions for improvements, please open an issue or submit a pull request on the GitHub repository for this package.

### License

This package is licensed under the [MIT License](https://opensource.org/licenses/MIT).
