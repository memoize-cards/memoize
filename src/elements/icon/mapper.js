import CaretLeft from './media/caretLeft.svg'
import CaretRight from './media/caretRight.svg'
import Live from './media/live.svg'
import TrendUp from './media/trendUp.svg'

/**
 * Retrieves the corresponding icon component based on the provided key.
 *
 * @param {string} key - The key representing the desired icon component.
 * @returns {import('path/to/icon/component')} The icon component corresponding to the provided key.
 */
function mapper (key) {
  return { CaretLeft, CaretRight, Live, TrendUp }[key]
}

export default mapper
