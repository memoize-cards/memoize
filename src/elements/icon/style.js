import css from '@standard/css'

const style = css`
  .icon {
    font-size: 24px;
  }

  .icon[small] {
    font-size: 16px;
  }

  .icon[large] {
    font-size: 32px;
  }
`

export default style
