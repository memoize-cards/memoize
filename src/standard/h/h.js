import Component from './component'
import Element from './element'

export default (nodeName, attrs, ...children) => {
  attrs ??= {}
  if (Element.is(nodeName)) return Element.create(nodeName, attrs, children)
  if (Component.is(nodeName)) return Component.execute(nodeName, attrs, children)
}
