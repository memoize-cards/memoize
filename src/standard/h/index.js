export { default } from './h'
export { hook as didMount } from './didMount'
export { hook as didUnmount } from './didUnmount'
export { hook as didUpdate } from './didUpdate'
export { default as Fragment } from './fragment'
export { default as paint } from './paint'
export { default as render } from './render'
export { default as repaint } from './repaint'
export { hook as willMount } from './willMount'
export { hook as willUnmount } from './willUnmount'
export { hook as willUpdate } from './willUpdate'
