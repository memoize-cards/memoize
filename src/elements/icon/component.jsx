import h from '@standard/h'

const icon = new Proxy({}, {
  get (_target, key) {
    return function component (props) {
      return (
        <span {...props} className={['material-symbols-outlined', props.className]}>{key.toLowerCase()}</span>
      )
    }
  }
})

export default icon
