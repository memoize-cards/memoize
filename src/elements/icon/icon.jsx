import h from '@standard/h'
import mapper from './mapper'
import style from './style'

/**
 * A utility function that returns a dynamically generated image element based on the provided key.
 * It creates an `img` element with the specified properties, using the `key` to determine the `src` attribute.
 * The `className` property is combined with the default `style.icon` class to style the icon.
 *
 * @param {string} key - The key to determine the source of the image.
 * @param {Object} props - Other attributes and properties to be applied to the `img` element.
 * @returns {Element} - The dynamically generated `img` element with the specified properties.
 */
const icon = new Proxy({}, {
  get (_target, key) {
    return (props) => (
      <img {...props} className={[style.icon, props.className]} src={mapper(key)} loading='lazy' />
    )
  }
})

export default icon
