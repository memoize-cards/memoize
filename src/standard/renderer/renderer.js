import { render } from '@standard/h'

async function renderer (element) {
  return (await element)[render.flow]?.()
}

export default renderer
