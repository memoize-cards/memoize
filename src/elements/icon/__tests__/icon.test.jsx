import h from '@standard/h'
import icon from '@elements/icon'
import mapper from '@elements/icon/mapper'
import renderer from '@standard/renderer'

jest.mock('@elements/icon/mapper', () => jest.fn().mockReturnValue('/url-mock.svg'))
jest.mock('@elements/icon/style', () => ({ icon: 'class-name-mock' }))

describe('@elements/icon', () => {
  beforeEach(() => {
    mapper.mockClear()
  })

  test('Should create an icon based on the tag name', () => {
    const tree = renderer(
      <icon.QCX />
    )

    expect(tree).toMatchSnapshot()
    expect(mapper).toHaveBeenCalledTimes(1)
    expect(mapper).toHaveBeenCalledWith('QCX')
  })

  test('Should pass the className to the icon', () => {
    const tree = renderer(
      <icon.QCX className='custom-class-name' />
    )

    expect(tree).toMatchSnapshot()
    expect(mapper).toHaveBeenCalledTimes(1)
    expect(mapper).toHaveBeenCalledWith('QCX')
  })
})
