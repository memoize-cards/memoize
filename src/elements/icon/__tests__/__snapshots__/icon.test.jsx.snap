// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@elements/icon Should create an icon based on the tag name 1`] = `
<img
  class="class-name-mock"
  loading="lazy"
  src="/url-mock.svg"
/>
`;

exports[`@elements/icon Should pass the className to the icon 1`] = `
<img
  class="class-name-mock custom-class-name"
  loading="lazy"
  src="/url-mock.svg"
/>
`;
